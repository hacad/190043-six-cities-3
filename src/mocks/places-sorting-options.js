const placesSortingOptions = [
  {caption: `Popular`, value: `popular`},
  {caption: `Price: low to high`, value: `price`, order: `ASC`},
  {caption: `Price: high to low`, value: `price`, order: `DESC`},
  {caption: `Top rated first`, value: `rating`, order: `DESC`}
];

const defaultSortingOptionItem = placesSortingOptions[0];

export {placesSortingOptions, defaultSortingOptionItem};
