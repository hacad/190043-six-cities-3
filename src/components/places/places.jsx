import React from "react";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import {ActionCreator} from "../../reducers/data/reducer.js";
import ReducerNames from "../../reducers/reducer-names.js";
import {getSortingOption} from "../../reducers/data/selectors.js";
import withActiveItem from "../../hocs/with-active-item/with-active-item.js";
import PlacesSortingPropType from "../../prop-types/places-sorting.js";
import PlacesSorting from "../places-sorting/places-sorting.jsx";
import PlacePropType from "../../prop-types/place.js";
import PlacesList from "../places-list/places-list.jsx";
import CitiesMap from "../cities-map/cities-map.jsx";
import CityPropType from "../../prop-types/city.js";
import {placesSortingOptions} from "../../mocks/places-sorting-options";

const PlacesSortingWrapped = withActiveItem(PlacesSorting, false, `Open`);
const PlacesListWrapped = withActiveItem(PlacesList);

function Places(props) {
  const {places, handleItemActivate, handleItemDeactivate, handleSortingChange,
    activeCity, activeOffer, selectedItem} = props;
  const offers = places.map((place) => {
    return {data: place.location};
  });
  const activeCitiesMapOffer = activeOffer
    ? {
      data: activeOffer.location,
      displaySettings: {
        icon: {
          iconUrl: `/img/pin-active.svg`
        }
      }
    }
    : undefined;

  return (
    <div className="cities__places-container container">
      <section className="cities__places places">
        <h2 className="visually-hidden">Places</h2>
        <b className="places__found">{offers.length} {offers.length === 1 ? `place` : `places`} to stay in {places[0].city.name}</b>
        <PlacesSortingWrapped
          activeCity={activeCity}
          selectedItem={selectedItem}
          items={placesSortingOptions}
          handleItemSelect={handleSortingChange}
        />
        <PlacesListWrapped
          places={places}
          handleClickCardHeader={() => {}}
          className="cities__places-list places__list tabs__content"
          handleActivatePlace={handleItemActivate}
          handleDeactivatePlace={handleItemDeactivate}
          articleTagClassNamePrefix="cities__place-card"
          divImageWrapperClassNamePrefix="cities__image-wrapper"
        />
      </section>
      <div className="cities__right-section">
        <CitiesMap city={activeCity} offers={offers} activeOffer={activeCitiesMapOffer} className="property__map map"/>
      </div>
    </div>
  );
}

Places.propTypes = {
  places: PropTypes.arrayOf(PlacePropType).isRequired,
  className: PropTypes.string.isRequired,
  handleClickCardHeader: PropTypes.func.isRequired,
  activeOffer: PlacePropType,
  handleItemActivate: PropTypes.func.isRequired,
  handleItemDeactivate: PropTypes.func.isRequired,
  handleSortingChange: PropTypes.func.isRequired,
  activeCity: CityPropType.isRequired,
  selectedItem: PlacesSortingPropType.isRequired,
};

const mapStateToProps = (state, ownProps) => {
  return Object.assign({}, ownProps, {
    selectedItem: getSortingOption(state),
    activeOffer: state[ReducerNames.DATA].activeOffer
  });
};

const mapDispatchToProps = (dispatch) => {
  return {
    handleSortingChange: (placeSortingOption) => {
      dispatch(ActionCreator.changeSorting(placeSortingOption));
    },
    handleItemActivate: (activeOffer) => {
      dispatch(ActionCreator.changeActiveOffer(activeOffer));
    },
    handleItemDeactivate: () => {
      dispatch(ActionCreator.changeActiveOffer(undefined));
    }
  };
};

export {Places};

export default connect(mapStateToProps, mapDispatchToProps)(Places);
