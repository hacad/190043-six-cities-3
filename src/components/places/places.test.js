import React from "react";
import renderer from "react-test-renderer";
import {Places} from "./places.jsx";
import {defaultSortingOptionItem} from "../../mocks/places-sorting-options.js";

jest.mock(`../place-card/place-card.jsx`, () => `<div />`);

it(`Places correctly renders after relaunch`, () => {
  // Arrange
  const places = [{
    id: 15,
    city: {
      name: `Hamburg`,
      location: {
        latitude: 53.552645,
        longitude: 9.966287,
        zoom: 10
      }
    },
    type: `apartment`,
    previewImage: `img/apartment-03.jpg`,
    isPremium: false,
    title: `Nice, cozy, warm big bed apartment`,
    price: 180,
    rating: 100,
    isFavorite: false,
    location: {
      latitude: 53.554221,
      longitude: 9.992588
    }
  }];

  const activeOffer = {
    id: 16,
    city: {
      name: `Paris`,
      location: {
        latitude: 48.856663,
        longitude: 2.351556,
        zoom: 10
      }
    },
    type: `apartment`,
    previewImage: `img/apartment-02.jpg`,
    isPremium: false,
    title: `Nice, cozy, warm big bed apartment`,
    price: 300,
    rating: 100,
    isFavorite: false,
    location: {
      latitude: 48.845933,
      longitude: 2.357839
    }
  };

  const placesComponent = renderer.create(
      <Places
        places={places}
        className="cities__places-list places__list tabs__content"
        handleClickCardHeader={jest.fn()}
        activeOffer={activeOffer}
        handleItemActivate={jest.fn()}
        handleItemDeactivate={jest.fn()}
        handleSortingChange={jest.fn()}
        activeCity={activeOffer.city}
        selectedItem={defaultSortingOptionItem}
      />
  );

  // Act & Assert
  expect(placesComponent).toMatchSnapshot();
});
