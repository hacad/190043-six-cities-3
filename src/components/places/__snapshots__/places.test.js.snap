// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Places correctly renders after relaunch 1`] = `
<div
  className="cities__places-container container"
>
  <section
    className="cities__places places"
  >
    <h2
      className="visually-hidden"
    >
      Places
    </h2>
    <b
      className="places__found"
    >
      1
       
      place
       to stay in 
      Hamburg
    </b>
    <form
      action="#"
      className="places__sorting"
      method="get"
    >
      <span
        className="places__sorting-caption"
      >
        Sort by
      </span>
      <span
        className="places__sorting-type"
        onClick={[Function]}
        tabIndex="0"
      >
        Popular
        <svg
          className="places__sorting-arrow"
          height="4"
          width="7"
        >
          <use
            xlinkHref="#icon-arrow-select"
          />
        </svg>
      </span>
      <ul
        className="places__options places__options--custom "
      >
        <li
          className="places__option places__option--active"
          onClick={[Function]}
          tabIndex={0}
        >
          Popular
        </li>
        <li
          className="places__option "
          onClick={[Function]}
          tabIndex={1}
        >
          Price: low to high
        </li>
        <li
          className="places__option "
          onClick={[Function]}
          tabIndex={2}
        >
          Price: high to low
        </li>
        <li
          className="places__option "
          onClick={[Function]}
          tabIndex={3}
        >
          Top rated first
        </li>
      </ul>
    </form>
    <div
      className="cities__places-list places__list tabs__content"
    >
      <<div />
        onActivate={[MockFunction]}
        onClickHeader={[Function]}
        onDeactivate={[MockFunction]}
        place={
          Object {
            "city": Object {
              "location": Object {
                "latitude": 53.552645,
                "longitude": 9.966287,
                "zoom": 10,
              },
              "name": "Hamburg",
            },
            "id": 15,
            "isFavorite": false,
            "isPremium": false,
            "location": Object {
              "latitude": 53.554221,
              "longitude": 9.992588,
            },
            "previewImage": "img/apartment-03.jpg",
            "price": 180,
            "rating": 100,
            "title": "Nice, cozy, warm big bed apartment",
            "type": "apartment",
          }
        }
      />
    </div>
  </section>
  <div
    className="cities__right-section"
  >
    <section
      className="property__map map"
      id="map"
    />
  </div>
</div>
`;
